File: ej5/ej5.pl
% # Ejercicio 2: Se pide armar el árbol familiar Real Inglés, usando la relación “esHijoDe”
% # ¿Quienes son primos?
% # BD:
esHijoDe(carlosIII, isabelII).
esHijoDe(ana, isabelII).
esHijoDe(andrew, isabelII).
esHijoDe(edward, isabelII).
esHijoDe(henry, carlosIII).
% # ¿Es cierto que la reina Isabel tiene hijos?	
esHijoDe(_,isabelII).
% # ¿Quién o quienes son hijos de la reina Isabel?
esHijoDe(Hijos,isabelII).
% # ¿Quiénes son los nietos del Carlos III?
esNieto(Nieto,Abuelo) :-
    esHijoDe(Nieto,Padre),
    esHijoDe(Padre,Abuelo),
    Nieto \= Abuelo.
% # ¿Es cierto que Archie es primo del príncipe Jorge?
esPrimo(Primo1,Primo2):-
    esHijoDe(Primo1, Padre1),
    esHijoDe(Primo2, Padre2),
    esHijoDe(Padre1, Abuelo),
    esHijoDe(Padre2, Abuelo),
    Primo1 \= Primo2, Padre1\=Padre2.

File: ej5/ej5.hs
-- Ejercicio 2: Se pide armar el árbol familiar Real Inglés, usando la relación “esHijoDe”
-- ¿Es cierto que Archie es primo del príncipe Jorge?
-- ¿Quienes son primos?
-- BD:
-- esHijoDe(carlosIII, isabelII).
-- esHijoDe(ana, isabelII).
-- esHijoDe(andrew, isabelII).
-- esHijoDe(edward, isabelII).
-- esHijoDe(henry, carlosII).
-- ……….
type Nombre = String

data Persona = Persona 
  { nombre :: Nombre
  , padres :: [Nombre] -- Lista con los nombres de sus padres
  } deriving (Show, Eq)

esHijoDe :: Persona -> Persona -> Bool
esHijoDe padre hijo = elem (nombre padre) (padres hijo)

-- ¿Es cierto que la reina Isabel tiene hijos?	
tieneHijos :: Persona -> [Persona] -> Bool
tieneHijos unaPersona personas = any (esHijoDe unaPersona) personas
-- Se puede usar lambda (\p -> esHijoDe persona p) pero se puede usar de la otra forma mas reducida

-- ¿Quién o quienes son hijos de la reina Isabel?
quienesSonHijos :: Persona -> [Persona] -> [Persona]
quienesSonHijos unaPersona personas = filter (esHijoDe unaPersona) personas

-- ¿Quiénes son los nietos del Carlos III?
quienesSonNietos :: Persona -> [Persona] -> [Persona]
quienesSonNietos abuelo personas = concatMap (\padre -> quienesSonHijos padre personas) (quienesSonHijos abuelo personas)
